### ISSUE TITLE
code specification: Should we use assertions in our code?

### ISSUE THREAD

**Issue by aniaan** [tiangolo/fastapi#2318]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
# save as main.py
from typing import Dict, List, Optional, Tuple

from fastapi import FastAPI, Query
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    title: str


@app.get("/items/")
def read_items(q: Tuple[Item, Item] = Query(None)):
    pass  # pragma: no cover
```

### Description

When I run `python main.py`, it throws an `AssertionError: Param: q can only be a request body, using Body(...) `, which is controlled by the code in https://github.com/tiangolo/fastapi/blob/77f7447ee8917697e702b7eca8d393e11dc1e365/fastapi/dependencies/utils.py#L331-L334, and I don't see any problem with throwing an exception.
What I want to discuss is whether or not we should use assert here, because when I run `python -O main.py`, the AssertionError is skipped here, and there is a bit of a problem here, whenever the exception is thrown because of a user error, we should come and notify the user, should we use `if statement` to detect user input errors, not assert, 

```python
if isinstance(field_info, params.Body):
    raise Exception(
        f"Param: {param_field.name} can only be a request body, using Body(...)"
    )
```
If possible, I am willing to submit pr to improve the code here

**Comment by Mause** [comment]
That error should only be thrown during development, so I don't think it's an issue that it doesn't get thrown in optimized mode

**Comment by BeanNan** [comment]
The semantics of assert is used for internal program self-testing, but here we are checking user input, and exceptions should be thrown at all times, using an if statement makes the program more robust and semantically clear.
As I browse through the code, I can see that some use assert and some use if to check, but the style is inconsistent.

Of course, this isn't a big deal, just discussing with everyone to make sure

**Comment by Mause** [comment]
It's not user input, it's a developer error, hence why this error only being thrown during development is reasonable

**Comment by BeanNan** [comment]
Because we're the framework side, the developers are our users, and the code they write is the input, so we should do the right exceptions to throw, and assert doesn't run in python-O. Isn't all we can do to be more thoughtful?

**Comment by ycd** [comment]
I agree with @Mause assert is raised only for developers, it does sanity-checking before you deploy your app to production, more specifically assertions raise `AssertionError`, which you really shouldn't catch. Also, they can be easily removed with the `-O` flag as you mentioned. So it's a win-win condition IMO. 

Just curious, are there any reasons to run an app with `-O` mode for local development?

**Comment by BeanNan** [comment]
Actually it's because of the -O option to skip checking that I have some doubts, I don't think the framer should allow exceptions at any time, we can use if statement to avoid this.

I haven't actually seen anyone using the -O option either, but that doesn't rule out not having it. I also agree with you that either is fine, but `if raise` can be made more complete. It's up to us to see if we need to change it.

**Comment by Mause** [comment]
It seems to me that your concern is purely hypothetical then? If someone has `-O` enabled during development, then they are probably doing things wrong

**Comment by BeanNan** [comment]
Yes, there are assumptions about this concern, and indeed the use of -O in development, itself, is their own problem.

**Comment by ihakh** [comment]
you can not use complex type as `Query`
sequences(list, set, tuple,...) of singular type (like int, float, str, bool, etc) is the only complex type that can be used in `Query`
the complex type should be used in `Body`
> If the parameter is also declared in the path, it will be used as a path parameter.
> If the parameter is of a singular type (like int, float, str, bool, etc) it will be interpreted as a query parameter.
> If the parameter is declared to be of the type of a Pydantic model, it will be interpreted as a request body.

**Comment by Kludex** [comment]
@ihakh it's the below issue, not this one :thinking:

**Comment by ihakh** [comment]
oops
the fastapi use these check to prevent app from deployment error
so you get these error during development and write a robust solid stable code which not fail during the deployment
if there a possible bad input from user you should add [validator](https://pydantic-docs.helpmanual.io/usage/validators/) to your code then return appropriate error to user

**Comment by BeanNan** [comment]
Thank you all for your answers, I have no more questions about this one.

**Comment by tiangolo** [comment]
Thanks for the help everyone! 🍰 

And thanks for coming back to close the issue @aniaan ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
